typedb/stream/response_collector.py

```diff
@@ -46,56 +46,71 @@ def new_queue(self, request_id: UUID):
     def get(self, request_id: UUID) -> Optional["ResponseCollector.Queue[R]"]:
         return self._collectors.get(request_id)
 
-    def close(self, error: Optional[RpcError]):
+    def remove(self, request_id: UUID):
+        with self._collectors_lock:
+            del self._collectors[request_id]
+
+    def close(self, error: Optional[TypeDBClientException]):
         with self._collectors_lock:
             for collector in self._collectors.values():
                 collector.close(error)
 
-    def drain_errors(self) -> [RpcError]:
+    def get_errors(self) -> [TypeDBClientException]:
         errors = []
         with self._collectors_lock:
             for collector in self._collectors.values():
-                errors.extend(collector.drain_errors())
+                error = collector.get_error()
+                if error is not None:
+                    errors.append(error)
         return errors
 
-
     class Queue(Generic[R]):
 
         def __init__(self):
             self._response_queue: queue.Queue[Union[Response[R], Done]] = queue.Queue()
+            self._error: TypeDBClientException = None
 
         def get(self, block: bool) -> R:
             response = self._response_queue.get(block=block)
-            if response.message:
+            if response.is_response():
                 return response.message
-            elif response.error:
-                raise TypeDBClientException.of_rpc(response.error)
-            else:
+            elif response.is_done() and self._error is None:
                 raise TypeDBClientException.of(TRANSACTION_CLOSED)
+            elif response.is_done() and self._error is not None:
+                raise TypeDBClientException.of(TRANSACTION_CLOSED_WITH_ERRORS, self._error)
+            else:
+                raise TypeDBClientException.of(ILLEGAL_STATE)
 
         def put(self, response: R):
             self._response_queue.put(Response(response))
 
-        def close(self, error: Optional[RpcError]):
-            self._response_queue.put(Done(error))
-
-        def drain_errors(self) -> [RpcError]:
-            errors = []
-            while not self._response_queue.empty():
-                response = self._response_queue.get(block = False)
-                if response.error:
-                    errors.append(response.error)
-            return errors
+        def close(self, error: Optional[TypeDBClientException]):
+            self._error = error
+            self._response_queue.put(Done())
 
+        def get_error(self) -> TypeDBClientException:
+            return self._error
 
 
 class Response(Generic[R]):
 
     def __init__(self, value: R):
         self.message = value
 
+    def is_response(self):
+        return True
```

Comment:
While not technically required (in either Python or TS) we did introduce a base class `QueueElement` for these two types of response (`Response` and `Done`) and given that we are now relying on the existence of a method `is_response` in each of them, we should definitely introduce a `QueueElement` base class in client-python too.